crates/circuit/src/slice.rs

```diff
@@ -0,0 +1,375 @@
+                        .ok_or(PySequenceIndexError::OutOfRange)?
+                };
+                Ok(SequenceIndex::Int(index))
+            }
+            PySequenceIndex::Slice(slice) => {
+                let indices = slice
+                    .indices(len as ::std::os::raw::c_long)
+                    .map_err(PySequenceIndexError::from)?;
+                if indices.step > 0 {
+                    Ok(SequenceIndex::PosRange {
+                        start: indices.start as usize,
+                        stop: indices.stop as usize,
+                        step: indices.step as usize,
+                    })
+                } else {
+                    Ok(SequenceIndex::NegRange {
+                        // `indices.start` can be negative if the collection length is 0.
+                        start: (indices.start >= 0).then_some(indices.start as usize),
+                        // `indices.stop` can be negative if the 0 index should be output.
+                        stop: (indices.stop >= 0).then_some(indices.stop as usize),
+                        step: indices.step.unsigned_abs(),
+                    })
+                }
+            }
+        }
+    }
+}
+
+/// Error type for problems encountered when calling methods on `PySequenceIndex`.
+#[derive(Error, Debug)]
+pub enum PySequenceIndexError {
+    #[error("index out of range")]
+    OutOfRange,
+    #[error(transparent)]
+    InnerPy(#[from] PyErr),
+}
+impl From<PySequenceIndexError> for PyErr {
+    fn from(value: PySequenceIndexError) -> PyErr {
+        match value {
+            PySequenceIndexError::OutOfRange => PyIndexError::new_err("index out of range"),
+            PySequenceIndexError::InnerPy(inner) => inner,
+        }
+    }
+}
+
+/// Rust-native version of a Python sequence-like indexer.
+///
+/// Typically this is constructed by a call to `PySequenceIndex::with_len`, which guarantees that
+/// all the indices will be in bounds for a collection of the given length.
+///
+/// This splits the positive- and negative-step versions of the slice in two so it can be translated
+/// more easily into static dispatch.  This type can be converted into several types of iterator.
+#[derive(Clone, Copy, Debug)]
+pub enum SequenceIndex {
+    Int(usize),
+    PosRange {
+        start: usize,
+        stop: usize,
+        step: usize,
+    },
+    NegRange {
+        start: Option<usize>,
+        stop: Option<usize>,
+        step: usize,
+    },
+}
+
+impl SequenceIndex {
+    /// The number of indices this refers to.
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Int(_) => 1,
+            Self::PosRange { start, stop, step } => {
+                let gap = stop.saturating_sub(*start);
+                gap / *step + (gap % *step != 0) as usize
+            }
+            Self::NegRange { start, stop, step } => 'arm: {
+                let Some(start) = start else { break 'arm 0 };
```

Comment:
Haha yeah, I meant the "match arm" but I didn't even think about that!

Yes, you can have two `None`, and in fact I think that's the _only_ time `indices.start` can be `None`.  Either if it weren't, the length would still always be zero because the start point isn't a valid index.